src/backend/latex/document/acmarticle.rs

```diff
@@ -0,0 +1,100 @@
+
+#[derive(Debug)]
+pub struct AcmArticle;
+
+#[rustfmt::skip]
+impl<'a> Backend<'a> for AcmArticle {
+    type Text = latex::TextGen;
+    type Latex = latex::LatexGen;
+    type FootnoteReference = latex::FootnoteReferenceGen;
+    type BiberReferences = latex::BiberReferencesGen;
+    type Url = latex::UrlGen;
+    type InterLink = latex::InterLinkGen;
+    type Image = latex::ImageGen;
+    type Label = latex::LabelGen;
+    type Pdf = latex::PdfGen;
+    type SoftBreak = latex::SoftBreakGen;
+    type HardBreak = latex::HardBreakGen;
+    type TaskListMarker = latex::TaskListMarkerGen;
+    type TableOfContents = latex::TableOfContentsGen;
+    type Bibliography = latex::BibliographyGen;
+    type ListOfTables = latex::ListOfTablesGen;
+    type ListOfFigures = latex::ListOfFiguresGen;
+    type ListOfListings = latex::ListOfListingsGen;
+    type Appendix = latex::AppendixGen;
+
+    type Paragraph = latex::ParagraphGen;
+    type Rule = latex::RuleGen;
+    type Header = latex::HeaderGen<'a>;
+    type BlockQuote = latex::BlockQuoteGen;
+    type CodeBlock = latex::CodeBlockGen;
+    type List = latex::ListGen;
+    type Enumerate = latex::EnumerateGen;
+    type Item = latex::ItemGen;
+    type FootnoteDefinition = latex::FootnoteDefinitionGen;
+    type UrlWithContent = latex::UrlWithContentGen<'a>;
+    type InterLinkWithContent = latex::InterLinkWithContentGen;
+    type HtmlBlock = latex::HtmlBlockGen;
+    type Figure = latex::FigureGen<'a>;
+
+    type TableFigure = latex::TableFigureGen<'a>;
+    type Table = latex::TableGen<'a>;
+    type TableHead = latex::TableHeadGen;
+    type TableRow = latex::TableRowGen;
+    type TableCell = latex::TableCellGen;
+
+    type InlineEmphasis = latex::InlineEmphasisGen;
+    type InlineStrong = latex::InlineStrongGen;
+    type InlineStrikethrough = latex::InlineStrikethroughGen;
+    type InlineCode = latex::InlineCodeGen;
+    type InlineMath = latex::InlineMathGen;
+
+    type Equation = latex::EquationGen<'a>;
+    type NumberedEquation = latex::NumberedEquationGen<'a>;
+    type Graphviz = latex::GraphvizGen<'a>;
+
+    fn new() -> Self {
+        AcmArticle
+    }
+
+    fn gen_preamble(&mut self, cfg: &Config, out: &mut impl Write, diagnostics: &Diagnostics<'a>) -> FatalResult<()> {
+        // TODO: itemizespacing
+        writeln!(out, "\\PassOptionsToPackage{{usenames,dvipsnames}}{{color}}")?;
+        writeln!(out, "\\PassOptionsToPackage{{pdfusetitle}}{{hyperref}}")?;
+        writeln!(out, "\\PassOptionsToPackage{{final}}{{microtype}}")?;
+        preamble::write_documentclass(cfg, out, "acmart", "nonacm,sigconf,natbib=false,pdfusetitle,")?;
+        preamble::write_packages(cfg, out)?;
+        preamble::write_fixes(cfg, out)?;
+
+        if let Some(abstract1) = &cfg.abstract1 {
+            writeln!(out, "\\begin{{abstract}}")?;
+            preamble::gen_abstract(abstract1.clone(), "abstract", &Arena::new(), AcmArticle, cfg, out, diagnostics)?;
```

Comment:
I find it weird that this creates a new `AcmArticle` and `Arena` instead of using `&mut self` and shared state. The former seems easier to address but for the latter the generator holding the usual Arena does not yet exist. It's also not a large problem.

Reply:
My idea here was that the abstract should be generated in isolation. Right now I don't see a reason why the abstracts should be treated context-sensitively, when they can just be treated context-free. Changing the callsites to pass `&mut self` instead of a new instance is easy enough, but that doesn't change anything as long as the abstracts can be context-free.

Re `Arena` the generator is already created, but reusing its arena means passing the arena to all `gen_preamble` functions. First, most of them don't require the arena (and for me it doesn't make sense to pass a reference to the arena into those functions generator-wise, that would only add more coupling between the generator and backend; also the generator is partially borrowed when calling `gen_preamble`, so we can't just pass a reference to the generator into those functions). Second, again, the abstract doesn't need to know anything about the rest of the document and vice versa. Reusing the global arena would mean that we keep the abstract string stored in memory forever, while creating a new arena will drop the abstract's source as soon as its finished generating.

I can do both of those changes, but I currently don't really see a reason why.

Reply:
What about the abstract is special that makes it distinct from other sections of the document? In fact, I wonder why the abstract appears in the pre-amble configuration at all. I should be able to cite with full power, put labels or reference them etc. As prior art in Latex documents the `abstract` is *contained* in the `document` and not entered before it.

Reply:
> I should be able to cite with full power, put labels or reference them etc.

That is a very good point I didn't really think about so far.

> As prior art in customary Latex classes the `abstract` is _contained_ in the `document` and not entered before it.

In our `thesis` (which uses `scrbook`) you're right, that the abstract is part of the `document`. However, it's still part of the `frontmatter` and in the TUM template should be placed before the tableofcontents. That means that we can't move to the `mainmatter` in the `gen_preamble` function.

https://github.com/oberien/heradoc/blob/2c6229f6ce58ab95a35d2d396aca270c6049ac13/src/backend/latex/document/thesis.rs#L100-L114

Additionally, for `acmart`, citing from the specification:

> The environment `abstract` must *precede* the `\maketitle` command. Again, this is different from the standard LATEX.  Putting `abstract` after `\maketitle` will trigger an error.
> – [<https://www.acm.org/binaries/content/assets/publications/consolidated-tex-template/acmart.pdf>, Page 17]

IMO those reasons make it pretty hard to include the `abstract` anywhere else than in the config referring to a different file. We could try some magic forkery where we parse the markdown twice (which we'll need to do anyways at some point due to footnotes), and in the first pass generate the abstract, which could be specified between `[abstract]` and `[endabstract]`.